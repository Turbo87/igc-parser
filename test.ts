import fs = require('fs');
import IGCParser = require('.');

describe('IGCParser', () => {
  let parser: IGCParser;
  let lenientParser: IGCParser;

  beforeEach(() => {
    parser = new IGCParser();
    lenientParser = new IGCParser({ lenient: true });
  });

  describe('parse()', () => {
    test('1G_77fv6m71.igc', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/1G_77fv6m71.igc`, 'utf8');
      let result = IGCParser.parse(content, { parseComments: true });

      expect(result.fixes.length).toEqual(4047);
      expect(result.dataRecords.length).toEqual(80);

      // reduce number of fixes to assert
      result.fixes = [
        result.fixes[0],
        result.fixes[1234],
        result.fixes[2042],
        result.fixes[4046],
      ];

      // reduce number of data records to assert
      result.dataRecords = [
        result.dataRecords[0],
        result.dataRecords[42],
        result.dataRecords[result.dataRecords.length - 1],
      ];

      expect(result).toMatchSnapshot();
    });

    test('654G6NG1.IGC', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/654G6NG1.IGC`, 'utf8');
      let result = IGCParser.parse(content);

      // reduce number of fixes to assert
      result.fixes = [
        result.fixes[0],
        result.fixes[123],
        result.fixes[420],
        result.fixes[3000],
        result.fixes[result.fixes.length - 1],
      ];

      expect(result).toMatchSnapshot();
    });

    test('2016-11-08-xcs-aaa-02.igc', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/2016-11-08-xcs-aaa-02.igc`, 'utf8');
      let result = IGCParser.parse(content);

      expect(result.fixes.length).toEqual(6752);

      // reduce number of fixes to assert
      result.fixes = [
        result.fixes[0],
        result.fixes[1234],
        result.fixes[2042],
        result.fixes[result.fixes.length - 1],
      ];

      expect(result).toMatchSnapshot();
    });

    test('20180427.igc', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/20180427.igc`, 'utf8');
      let result = IGCParser.parse(content);

      expect(result.fixes.length).toEqual(1831);

      // reduce number of fixes to assert
      result.fixes = [
        result.fixes[0],
        result.fixes[234],
        result.fixes[1777],
        result.fixes[result.fixes.length - 1],
      ];

      expect(result).toMatchSnapshot();
    });

    test('20211015.igc', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/20211015.igc`, 'utf8');
      let result = IGCParser.parse(content);

      expect(result.fixes.length).toEqual(4886);
      expect(result.site).toEqual("Riederalp");

      // reduce number of fixes to assert
      result.fixes = [
        result.fixes[0],
        result.fixes[234],
        result.fixes[1777],
        result.fixes[result.fixes.length - 1],
      ];

      expect(result).toMatchSnapshot();
    });

    /* Test timezone parsing */
    test('20241007TZN.igc', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/20241007TZN.igc`, 'utf8');
      let result = IGCParser.parse(content);

      expect(result.timezone).toEqual(5.5);
    });

    test('MD_85ugkjj1.IGC', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/MD_85ugkjj1.IGC`, 'utf8');
      let result = IGCParser.parse(content, { lenient: true });

      expect(result.fixes.length).toEqual(8924);
      expect(result.errors).toMatchSnapshot();
    });

    test('lad_lod_extensions.igc', () => {
      let content = fs.readFileSync(`${__dirname}/fixtures/lad_lod_extensions.igc`, 'utf8');
      let result = IGCParser.parse(content, { lenient: true });

      expect(result.fixes.length).toEqual(424);

      expect(result.fixes[0].latitude).toBeCloseTo(44.968046666666666, 10);
      expect(result.fixes[0].longitude).toBeCloseTo(5.833138333333333, 10);

      expect(result.errors).toEqual([]);
    });

    it('throws if HFDTE is missing', () => {
      let lines = [
        'ALXV6M7FLIGHT:1',
        'HFGIDGLIDERID:D-2019',
      ];

      expect(() => IGCParser.parse(lines.join('\n'))).toThrowErrorMatchingSnapshot();
    });

    it('handles UTC midnight correctly', () => {
      let result = IGCParser.parse([
        'ALXV6M7',
        'HFDTE150717',
        'B1926130000000N00000000EA0000000000',
        'B2300210000000N00000000EA0000000000',
        'B0001550000000N00000000EA0000000000',
        'B2359590000000N00000000EA0000000000',
        'B0242420000000N00000000EA0000000000',
        'B0659000000000N00000000EA0000000000',
        'B2200000000000N00000000EA0000000000',
        'B0100000000000N00000000EA0000000000',
      ].join('\n'));

      let fixTimestamps = result.fixes.map(fix => (new Date(fix.timestamp)).toISOString());
      expect(fixTimestamps).toEqual([
        '2017-07-15T19:26:13.000Z',
        '2017-07-15T23:00:21.000Z',
        '2017-07-16T00:01:55.000Z',
        '2017-07-15T23:59:59.000Z',
        '2017-07-16T02:42:42.000Z',
        '2017-07-16T06:59:00.000Z',
        '2017-07-16T22:00:00.000Z',
        '2017-07-17T01:00:00.000Z',
      ]);
    });

    it('handles GPS Datum FAI Sporting Code 2021 0.9.0 - 3.2.3 & 3.4', () => {
      // Example data as of 3.4 - see fai.org
      let result = IGCParser.parse([
        'AVXX00026',
        'HFDTEDATE:160816',
        'HFPLTPILOTINCHARGE:Bloggs Bill',
        'HFCM2CREW2:NIL',
        'HFGTYGLIDERTYPE:NKN',
        'HFGIDGLIDERID:NKN',
        'HFDTMGPSDATUM:WGS84',
        'HFRFWFIRMWAREVERSION:0.2-alpha',
        'HFRHWHARDWAREVERSION:1.0',
        'HFFTYFRTYPE:Zebra Instruments,Proto 1',
        'HFGPSRECEIVER:UBLOX,NEO7,56,50000',
        'HFPRSPRESSALTSENSOR:MEAS,MS5611,25907',
        'HFALGALTGPS:GEO',
        'HFALPALTPRESSURE:ISA',
        'B1153555536248N00339528WA0050200475',
        'B1154005536249N00339528WA0050300476',
        'B1154105536249N00339528WA0050300477',
        'B1154155536248N00339528WA0050300476',
        'G5734B6437B7796F96460F5D8AAC8FD4F',
        'G0B6401B0E19216179A25DAE23CD0487F',
      ].join('\n'));

      expect(result.geoDatum).toEqual('WGS84');
      expect(result.geoDatumAlgorithm).toEqual('GEO');
      expect(result.geoPressureAlgorithm).toEqual('ISA');
    });
  });

  describeMethod('parseARecord', (test) => {
    test.pass('AFLA1BG', {
      manufacturer: 'Flarm',
      loggerId: '1BG',
      numFlight: null,
      additionalData: null,
    });

    test.pass('AMMMNNN:TEXT STRING', {
      manufacturer: 'MMM',
      loggerId: 'NNN',
      numFlight: null,
      additionalData: 'TEXT STRING',
    });

    test.pass('ALXV6M7FLIGHT:1', {
      manufacturer: 'LXNAV',
      loggerId: '6M7',
      numFlight: 1,
      additionalData: null,
    });

    test.pass('ALXNNE2FLIGHT:42', {
      manufacturer: 'LX Navigation',
      loggerId: 'NE2',
      numFlight: 42,
      additionalData: null,
    });

    test.pass('AXYZ123456:foo', {
      manufacturer: 'XYZ',
      loggerId: '123456',
      numFlight: null,
      additionalData: 'foo',
    });

    test.pass('AXCT XCTrack 0.6.0 on samsung GT-S6500 2.3.6', {
      manufacturer: 'XCTrack',
      loggerId: null,
      numFlight: null,
      additionalData: 'XCTrack 0.6.0 on samsung GT-S6500 2.3.6',
    });

    test.pass('AXGD Flymaster F1, V1.21, S/N 3556', {
      manufacturer: 'GpsDump',
      loggerId: null,
      numFlight: null,
      additionalData: 'Flymaster F1, V1.21, S/N 3556',
    });

    test.pass('AXSR', {
      manufacturer: 'XSR',
      loggerId: null,
      numFlight: null,
      additionalData: null,
    });

    test.fail('');
    test.fail('HFDTE');
    test.fail('HFDTEXXXXXX');
    test.fail('HFDTE1234?6');
  });

  describeMethod('parseDateHeader', (test) => {
    test.pass('HFDTE010180', { date: '1980-01-01', numFlight: null });
    test.pass('HFDTE150717', { date: '2017-07-15', numFlight: null });
    test.pass('HFDTE15071799', { date: '2017-07-15', numFlight: 99 });
    test.pass('HFDTE311279', { date: '2079-12-31', numFlight: null });
    test.pass('HFDTEDATE:150717', { date: '2017-07-15', numFlight: null });
    test.pass('HFDTEDATE:15071702', { date: '2017-07-15', numFlight: 2 });
    test.pass('HFDTEDATE:150717,01', { date: '2017-07-15', numFlight: 1 });

    test.fail('');
    test.fail('HFDTE');
    test.fail('HFDTEXXXXXX');
    test.fail('HFDTE1234?6');
  });

  describeMethod('parsePilot', (test) => {
    test.pass('HFPLTPILOT:', '');
    test.pass('HFPLTPilotincharge:Mike Havener', 'Mike Havener');
    test.pass('HFPLTPILOT:BEN BRAND', 'BEN BRAND');
    test.pass('HFPLTPILOT:foo:bar', 'foo:bar');
    test.pass('HFPLTPILOTINCHARGE:YUTA_ISHIGURO', 'YUTA ISHIGURO');
    test.pass('HFPLT Pilot             :Gisela Weinreich  ', 'Gisela Weinreich');
    test.pass('HFPLT:John Doe', 'John Doe');
    test.pass('HOPLTJohn Doe', 'John Doe');

    test.fail('');
    test.fail('HXPLT:John Doe');
    test.fail('HFPLX:John Doe');
  });

  describeMethod('parseCopilot', (test) => {
    test.pass('HFCM2CREW2:', '');
    test.pass('HFCM2Crew 2:Mike Havener', 'Mike Havener');
    test.pass('HFCM2CREW2:BEN BRAND', 'BEN BRAND');
    test.pass('HFCM2CREW2:foo:bar', 'foo:bar');
    test.pass('HFCM2CREW2:YUTA_ISHIGURO', 'YUTA ISHIGURO');
    test.pass('HFCM2 Copilot           :Gisela Weinreich  ', 'Gisela Weinreich');
    test.pass('HFCM2:John Doe', 'John Doe');
    test.pass('HOCM2John Doe', 'John Doe');
    test.pass('HPCM2:John Doe', 'John Doe');

    test.fail('');
    test.fail('HXCM2:John Doe');
    test.fail('HFCM1:John Doe');

    test.pass('HXCM2:John Doe', 'John Doe', { lenient: true });
  });

  describeMethod('parseTask', (test) => {
    test.pass('C150717085720000000000204', {
      comment: null,
      declarationDate: '2017-07-15',
      declarationTime: '08:57:20',
      declarationTimestamp: 1500109040000,
      flightDate: null,
      numTurnpoints: 4,
      points: [],
      taskNumber: 2,
    });

    test.pass('C1707171703131707170001-2', {
      comment: null,
      declarationDate: '2017-07-17',
      declarationTime: '17:03:13',
      declarationTimestamp: 1500310993000,
      flightDate: '2017-07-17',
      numTurnpoints: -2,
      points: [],
      taskNumber: 1,
    });

    test.fail('');
    test.fail('C17071717031317');
  });

  describeMethod('parseBRecord', (test) => {
    beforeEach(() => {
      parser['_result'].date = '2017-02-03';
    });

    test.pass('B1026555103888N00703115EA0065700751', {
      timestamp: 1486117615000,
      time: '10:26:55',
      latitude: 51.0648,
      longitude: 7.051916666666667,
      valid: true,
      pressureAltitude: 657,
      gpsAltitude: 751,
      extensions: {},
      fixAccuracy: null,
      enl: null,
    });

    test.pass('B0000005103888N00703115EV0000000000', {
      timestamp: 1486080000000,
      time: '00:00:00',
      latitude: 51.0648,
      longitude: 7.051916666666667,
      valid: false,
      pressureAltitude: null,
      gpsAltitude: null,
      extensions: {},
      fixAccuracy: null,
      enl: null,
    });

    test.fail('');
    test.fail('B10a6555103888N00703115EA0065700751');
    test.fail('B1026555103888X00703115EA0065700751');
    test.fail('B102655510Z888N00703115EA0065700751');
    test.fail('B1026555103888N00703115NA0065700751');
  });

  describe('parseBRecord()', () => {
    it('throws for missing "dateHeader"', () => {
      expect(() => parser['parseBRecord']('B1026555103888N00703115EA0065700751')).toThrowErrorMatchingSnapshot();
    });
  });

  describeMethod('parseIJRecord', (test) => {
    test.pass('I013638FXA', [
      { code: 'FXA', start: 35, length: 3 },
    ]);

    test.pass('J043638FXA3940SIU4143ENL4446MOP', [
      { code: 'FXA', start: 35, length: 3 },
      { code: 'SIU', start: 38, length: 2 },
      { code: 'ENL', start: 40, length: 3 },
      { code: 'MOP', start: 43, length: 3 },
    ]);

    test.pass('I083638FXA3941ENL4246TAS4751GSP5254TRT5559VAT6063OAT6467ACZ', [
      { code: 'FXA', start: 35, length: 3 },
      { code: 'ENL', start: 38, length: 3 },
      { code: 'TAS', start: 41, length: 5 },
      { code: 'GSP', start: 46, length: 5 },
      { code: 'TRT', start: 51, length: 3 },
      { code: 'VAT', start: 54, length: 5 },
      { code: 'OAT', start: 59, length: 4 },
      { code: 'ACZ', start: 63, length: 4 },
    ]);

    test.pass('I023636LAD3737LOD', [
      { code: 'LAD', start: 35, length: 1 },
      { code: 'LOD', start: 36, length: 1 },
    ]);

    test.fail('');
    test.fail('I023638FXA');
    test.fail('I0136FXA38');
  });

  describeMethod('parseGeoDatum', (test) => {
    test.pass('HFDTMGPSDATUM:WGS84', 'WGS84');
    test.pass('HFDTM100GPSDATUM:WGS-1984', 'WGS-1984');
    test.pass('HFDTM100GPSDATUM:WGS-84', 'WGS-84');

    test.fail('');
  });

  describeMethod('parseGeoDatumAlgorithm', (test) => {
    test.pass('HFALGALTGPS:GEO', 'GEO');

    test.fail('');
  });

  describeMethod('parseGeoPressureAlgorithm', (test) => {
    test.pass('HFALPALTPRESSURE:ISA', 'ISA');

    test.fail('');
  });

  describeMethod('parseLRecord', (test) => {
    test.pass('LXMP 2 Trackpoints removed by user',
      { code: 'XMP', message: "2 Trackpoints removed by user" },
    );

    test.fail('');
    test.fail('LXP code to short');
  });

  // Test Suite Generator

  function describeMethod(methodName: string, cb: (test: any) => void) {
    describe(`${methodName}()`, () => {
      cb({
        pass(input: string, expected: any, options: { lenient?: boolean } = {}) {
          let testName = input ? input : '[empty]';
          if (typeof expected === 'string') {
            testName += ` -> ${expected ? `'${expected}'` : '[empty]'}`;
          }

          test(testName, () => {
            let _parser = options.lenient ? lenientParser : parser as any;
            expect(_parser[methodName](input)).toEqual(expected);
          });
        },

        fail(input: string, options: { lenient?: boolean } = {}) {
          test(`${input ? input : '[empty]'} -> 💥 `, () => {
            let _parser = options.lenient ? lenientParser : parser as any;
            expect(() => _parser[methodName](input)).toThrowErrorMatchingSnapshot();
          });
        },
      });
    });
  }
});
