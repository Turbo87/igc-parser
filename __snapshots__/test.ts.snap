// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IGCParser parse() 1G_77fv6m71.igc 1`] = `
Object {
  "callsign": "1G",
  "competitionClass": "Club",
  "copilot": null,
  "dataRecords": Array [
    Object {
      "extensions": Object {
        "WDI": "320",
        "WVE": "0071",
      },
      "time": "10:18:59",
      "timestamp": 1500113939000,
    },
    Object {
      "extensions": Object {
        "WDI": "880",
        "WVE": "1117",
      },
      "time": "12:37:02",
      "timestamp": 1500122222000,
    },
    Object {
      "extensions": Object {
        "WDI": "840",
        "WVE": "1002",
      },
      "time": "14:36:18",
      "timestamp": 1500129378000,
    },
  ],
  "date": "2017-07-15",
  "firmwareVersion": "6.0rc6",
  "fixes": Array [
    Object {
      "enl": 0.04,
      "extensions": Object {
        "ACZ": "100",
        "ENL": "040",
        "FXA": "060",
        "GSP": "00051",
        "OAT": "2400",
        "TAS": "00000",
        "TRT": "650",
        "VAT": "00010",
      },
      "fixAccuracy": 60,
      "gpsAltitude": 49,
      "latitude": 51.0107,
      "longitude": 7.010066666666667,
      "pressureAltitude": -42,
      "time": "10:18:26",
      "timestamp": 1500113906000,
      "valid": true,
    },
    Object {
      "enl": 0.041,
      "extensions": Object {
        "ACZ": "090",
        "ENL": "041",
        "FXA": "070",
        "GSP": "06662",
        "OAT": "2010",
        "TAS": "24341",
        "TRT": "60-",
        "VAT": "01290",
      },
      "fixAccuracy": 70,
      "gpsAltitude": 873,
      "latitude": 50.84336666666667,
      "longitude": 6.4120333333333335,
      "pressureAltitude": 773,
      "time": "11:37:45",
      "timestamp": 1500118665000,
      "valid": true,
    },
    Object {
      "enl": 0.04,
      "extensions": Object {
        "ACZ": "120",
        "ENL": "040",
        "FXA": "070",
        "GSP": "90033",
        "OAT": "2190",
        "TAS": "89120",
        "TRT": "54-",
        "VAT": "00740",
      },
      "fixAccuracy": 70,
      "gpsAltitude": 478,
      "latitude": 50.60658333333333,
      "longitude": 6.79495,
      "pressureAltitude": 378,
      "time": "12:28:38",
      "timestamp": 1500121718000,
      "valid": true,
    },
    Object {
      "enl": 0.04,
      "extensions": Object {
        "ACZ": "090",
        "ENL": "040",
        "FXA": "060",
        "GSP": "00153",
        "OAT": "2610",
        "TAS": "00000",
        "TRT": "30-",
        "VAT": "00010",
      },
      "fixAccuracy": 60,
      "gpsAltitude": 50,
      "latitude": 51.0137,
      "longitude": 7.007866666666667,
      "pressureAltitude": -40,
      "time": "14:39:10",
      "timestamp": 1500129550000,
      "valid": true,
    },
  ],
  "gliderType": "ASW 19",
  "hardwareVersion": "23",
  "loggerId": "6M7",
  "loggerManufacturer": "LXNAV",
  "loggerType": "LXNAV,LX8080",
  "numFlight": 1,
  "pilot": "Florian Graf",
  "registration": "D-2019",
  "security": "440FA49446EB4CDA7E1CB4B7D1F33A33F421E7A92656F2AEAE0A18356531A5B1E7C45C72F5C819E94A5DB32640019DC09BAB2EE64BE9FD1B558B9538308598C71F34C38E008D1AB9B859C762747F6B862B52D377C1D1C4D2BFDE801B63B9013AB53F4DF1A5DA6B1D8BCEF04ECF0D54D5895FF1C00DF131F7030241CB7E9C3FAEC1B8D",
  "task": Object {
    "comment": null,
    "declarationDate": "2017-07-15",
    "declarationTime": "08:57:20",
    "declarationTimestamp": 1500109040000,
    "flightDate": null,
    "numTurnpoints": 4,
    "points": Array [
      Object {
        "latitude": 0,
        "longitude": 0,
        "name": null,
      },
      Object {
        "latitude": 51.14138333333333,
        "longitude": 6.985283333333333,
        "name": "006Langenfeld-Wiescheid",
      },
      Object {
        "latitude": 50.82416666666667,
        "longitude": 6.18695,
        "name": "009Aachen-Merzbrueck",
      },
      Object {
        "latitude": 50.55971666666667,
        "longitude": 7.1375,
        "name": "019Bad Neuenahr",
      },
      Object {
        "latitude": 50.88888333333333,
        "longitude": 6.491666666666666,
        "name": "058Hambach Sud",
      },
      Object {
        "latitude": 51.17695,
        "longitude": 7.2,
        "name": "110Remscheid Bhf",
      },
      Object {
        "latitude": 51.09778333333333,
        "longitude": 7.03695,
        "name": "002Zielkreis",
      },
      Object {
        "latitude": 0,
        "longitude": 0,
        "name": null,
      },
    ],
    "taskNumber": 2,
  },
}
`;

exports[`IGCParser parse() 654G6NG1.IGC 1`] = `
Object {
  "callsign": "TH",
  "competitionClass": "Club",
  "copilot": "",
  "dataRecords": Array [],
  "date": "2016-05-04",
  "firmwareVersion": "Flarm-IGC06.01",
  "fixes": Array [
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "020",
        "SIU": "8",
      },
      "fixAccuracy": 20,
      "gpsAltitude": 230,
      "latitude": 50.82191666666667,
      "longitude": 6.18165,
      "pressureAltitude": 48,
      "time": "08:10:42",
      "timestamp": 1462349442000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "030",
        "SIU": "9",
      },
      "fixAccuracy": 30,
      "gpsAltitude": 1223,
      "latitude": 50.803333333333335,
      "longitude": 6.216966666666667,
      "pressureAltitude": 1067,
      "time": "08:19:54",
      "timestamp": 1462349994000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "021",
        "SIU": "0",
      },
      "fixAccuracy": 21,
      "gpsAltitude": 1013,
      "latitude": 50.6964,
      "longitude": 6.4348,
      "pressureAltitude": 846,
      "time": "08:39:42",
      "timestamp": 1462351182000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "030",
        "SIU": "9",
      },
      "fixAccuracy": 30,
      "gpsAltitude": 1857,
      "latitude": 49.21015,
      "longitude": 7.758483333333333,
      "pressureAltitude": 1718,
      "time": "11:31:42",
      "timestamp": 1462361502000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "031",
        "SIU": "0",
      },
      "fixAccuracy": 31,
      "gpsAltitude": 238,
      "latitude": 50.82215,
      "longitude": 6.189266666666667,
      "pressureAltitude": 72,
      "time": "19:04:54",
      "timestamp": 1462388694000,
      "valid": true,
    },
  ],
  "gliderType": "ASG-29E (18m)",
  "hardwareVersion": "Flarm-IGC06",
  "loggerId": "6NG",
  "loggerManufacturer": "Flarm",
  "loggerType": "Flarm-IGC",
  "numFlight": null,
  "pilot": "",
  "registration": "D-KCSS",
  "security": "60DC059E2D2F6CAD2E889224E355DBDDB805CAB100000639B49FC5F280A292C990F554789F12381380720000",
  "task": Object {
    "comment": "Task",
    "declarationDate": "2016-05-04",
    "declarationTime": "07:29:24",
    "declarationTimestamp": 1462346964000,
    "flightDate": null,
    "numTurnpoints": 2,
    "points": Array [
      Object {
        "latitude": 0,
        "longitude": 0,
        "name": "T",
      },
      Object {
        "latitude": 50.82305,
        "longitude": 6.186383333333334,
        "name": "1",
      },
      Object {
        "latitude": 49.052216666666666,
        "longitude": 7.956383333333333,
        "name": "2",
      },
      Object {
        "latitude": 50.63028333333333,
        "longitude": 10.158616666666667,
        "name": "3",
      },
      Object {
        "latitude": 50.82305,
        "longitude": 6.186383333333334,
        "name": "4",
      },
      Object {
        "latitude": 0,
        "longitude": 0,
        "name": "L",
      },
    ],
    "taskNumber": null,
  },
}
`;

exports[`IGCParser parse() 2016-11-08-xcs-aaa-02.igc 1`] = `
Object {
  "callsign": "",
  "competitionClass": null,
  "copilot": null,
  "dataRecords": Array [],
  "date": "2016-11-08",
  "firmwareVersion": null,
  "fixes": Array [
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "000",
        "SIU": "0",
      },
      "fixAccuracy": 0,
      "gpsAltitude": 423,
      "latitude": -44.48753333333333,
      "longitude": 169.98871666666668,
      "pressureAltitude": 468,
      "time": "22:43:17",
      "timestamp": 1478644997000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "000",
        "SIU": "0",
      },
      "fixAccuracy": 0,
      "gpsAltitude": 3608,
      "latitude": -44.60811666666667,
      "longitude": 169.91131666666666,
      "pressureAltitude": 3609,
      "time": "23:40:21",
      "timestamp": 1478648421000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "000",
        "SIU": "0",
      },
      "fixAccuracy": 0,
      "gpsAltitude": 2936,
      "latitude": -43.92655,
      "longitude": 169.8921,
      "pressureAltitude": 2935,
      "time": "00:45:32",
      "timestamp": 1478652332000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {
        "FXA": "000",
        "SIU": "0",
      },
      "fixAccuracy": 0,
      "gpsAltitude": 426,
      "latitude": -44.48518333333333,
      "longitude": 169.98096666666666,
      "pressureAltitude": 474,
      "time": "04:43:01",
      "timestamp": 1478666581000,
      "valid": true,
    },
  ],
  "gliderType": "DUO DISCUS",
  "hardwareVersion": null,
  "loggerId": "AAA",
  "loggerManufacturer": "XCSoar",
  "loggerType": "XCSOAR,XCSOAR Android 6.8.6 Jul 22 2016",
  "numFlight": null,
  "pilot": "ARNE MARTIN GUETTLER",
  "registration": "DUO",
  "security": "34201ec0e3908467ce7fe0fb31f9fe7556578b568a7474e300f763e6afa607f5cdd21b4e2a65ff9f23a64632e2ae11d1d3c5f443c78addd3da2a0e286ad79cee",
  "task": Object {
    "comment": null,
    "declarationDate": "2016-11-08",
    "declarationTime": "22:43:17",
    "declarationTimestamp": 1478644997000,
    "flightDate": null,
    "numTurnpoints": 2,
    "points": Array [
      Object {
        "latitude": 0,
        "longitude": 0,
        "name": "TAKEOFF",
      },
      Object {
        "latitude": -43.766666666666666,
        "longitude": 170.12833333333333,
        "name": "L235-MT COOK",
      },
      Object {
        "latitude": -45.51166666666666,
        "longitude": 169.31666666666666,
        "name": "L530 ROXBURGH",
      },
      Object {
        "latitude": -43.766666666666666,
        "longitude": 170.12833333333333,
        "name": "L235-MT COOK",
      },
      Object {
        "latitude": -44.48383333333334,
        "longitude": 169.97833333333332,
        "name": "001-OMARAMA",
      },
      Object {
        "latitude": 0,
        "longitude": 0,
        "name": "LANDING",
      },
    ],
    "taskNumber": null,
  },
}
`;

exports[`IGCParser parse() 20180427.igc 1`] = `
Object {
  "callsign": "86",
  "competitionClass": "",
  "copilot": null,
  "dataRecords": Array [],
  "date": "2018-04-27",
  "firmwareVersion": null,
  "fixes": Array [
    Object {
      "enl": null,
      "extensions": Object {},
      "fixAccuracy": null,
      "gpsAltitude": 583,
      "latitude": 45.9636,
      "longitude": 13.723516666666667,
      "pressureAltitude": 596,
      "time": "13:35:15",
      "timestamp": 1524836115000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {},
      "fixAccuracy": null,
      "gpsAltitude": 570,
      "latitude": 45.93281666666667,
      "longitude": 13.739566666666667,
      "pressureAltitude": 582,
      "time": "13:54:01",
      "timestamp": 1524837241000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {},
      "fixAccuracy": null,
      "gpsAltitude": 261,
      "latitude": 45.95603333333333,
      "longitude": 13.699766666666667,
      "pressureAltitude": 289,
      "time": "15:59:56",
      "timestamp": 1524844796000,
      "valid": true,
    },
    Object {
      "enl": null,
      "extensions": Object {},
      "fixAccuracy": null,
      "gpsAltitude": 57,
      "latitude": 45.94753333333333,
      "longitude": 13.712033333333334,
      "pressureAltitude": 97,
      "time": "16:03:25",
      "timestamp": 1524845005000,
      "valid": true,
    },
  ],
  "gliderType": "",
  "hardwareVersion": null,
  "loggerId": "000",
  "loggerManufacturer": "GpsDump",
  "loggerType": null,
  "numFlight": 1,
  "pilot": "Stefan Vyparina",
  "registration": "",
  "security": null,
  "task": null,
}
`;

exports[`IGCParser parse() throws if HFDTE is missing 1`] = `"Missing HFDTE record"`;

exports[`IGCParser parseARecord() [empty] -> 💥  1`] = `"Invalid A record at line 0: "`;

exports[`IGCParser parseARecord() HFDTE -> 💥  1`] = `"Invalid A record at line 0: HFDTE"`;

exports[`IGCParser parseARecord() HFDTE1234?6 -> 💥  1`] = `"Invalid A record at line 0: HFDTE1234?6"`;

exports[`IGCParser parseARecord() HFDTEXXXXXX -> 💥  1`] = `"Invalid A record at line 0: HFDTEXXXXXX"`;

exports[`IGCParser parseBRecord() [empty] -> 💥  1`] = `"Invalid B record at line 0: "`;

exports[`IGCParser parseBRecord() B10a6555103888N00703115EA0065700751 -> 💥  1`] = `"Invalid B record at line 0: B10a6555103888N00703115EA0065700751"`;

exports[`IGCParser parseBRecord() B102655510Z888N00703115EA0065700751 -> 💥  1`] = `"Invalid B record at line 0: B102655510Z888N00703115EA0065700751"`;

exports[`IGCParser parseBRecord() B1026555103888N00703115NA0065700751 -> 💥  1`] = `"Invalid B record at line 0: B1026555103888N00703115NA0065700751"`;

exports[`IGCParser parseBRecord() B1026555103888X00703115EA0065700751 -> 💥  1`] = `"Invalid B record at line 0: B1026555103888X00703115EA0065700751"`;

exports[`IGCParser parseBRecord() throws for missing "dateHeader" 1`] = `"Missing HFDTE record before first B record"`;

exports[`IGCParser parseCopilot() [empty] -> 💥  1`] = `"Invalid CM2 header at line 0: "`;

exports[`IGCParser parseCopilot() HFCM1:John Doe -> 💥  1`] = `"Invalid CM2 header at line 0: HFCM1:John Doe"`;

exports[`IGCParser parseCopilot() HXCM2:John Doe -> 💥  1`] = `"Invalid CM2 header at line 0: HXCM2:John Doe"`;

exports[`IGCParser parseDateHeader() [empty] -> 💥  1`] = `"Invalid DTE header at line 0: "`;

exports[`IGCParser parseDateHeader() HFDTE -> 💥  1`] = `"Invalid DTE header at line 0: HFDTE"`;

exports[`IGCParser parseDateHeader() HFDTE1234?6 -> 💥  1`] = `"Invalid DTE header at line 0: HFDTE1234?6"`;

exports[`IGCParser parseDateHeader() HFDTEXXXXXX -> 💥  1`] = `"Invalid DTE header at line 0: HFDTEXXXXXX"`;

exports[`IGCParser parseIJRecord() [empty] -> 💥  1`] = `"Invalid undefined record at line 0: "`;

exports[`IGCParser parseIJRecord() I0136FXA38 -> 💥  1`] = `"Invalid I record at line 0: I0136FXA38"`;

exports[`IGCParser parseIJRecord() I023638FXA -> 💥  1`] = `"Invalid I record at line 0: I023638FXA"`;

exports[`IGCParser parsePilot() [empty] -> 💥  1`] = `"Invalid PLT header at line 0: "`;

exports[`IGCParser parsePilot() HFPLX:John Doe -> 💥  1`] = `"Invalid PLT header at line 0: HFPLX:John Doe"`;

exports[`IGCParser parsePilot() HXPLT:John Doe -> 💥  1`] = `"Invalid PLT header at line 0: HXPLT:John Doe"`;
